File: F# Workspace/05_Unions/05_Unions/shapes_calc.fs
(*
 * shapes_calc.fs
 * By Antonio F. Huertas
 * This program displays the data, including the area and perimeter,
 * of a list of shapes.
 *)

// Defines a record type for a point.
type point = { x : float; y : float }

// Defines a discriminated union type for a color.
type color = White | Red | Green | Blue | Yellow | Black

// Define a discriminated union for a shape.
type Shape =
        // A circle has color, location and radius
    | Circle of color * point * float
        // A rectangle has color, location, height and width
    | Rectangle of color * point * float * float

// pi : float
// Defines the value of pi.
let pi = 3.14159

// area_shape : Shape -> float
// Returns the area of a shape.
let area_shape = function
    | Circle(_, _, rad) -> pi * rad * rad
    | Rectangle(_, _, hgt, wdt) -> hgt * wdt

// show_shape : Shape -> unit
// Returns the perimeter of a shape.
let perim_shape = function
    | Circle(_, _, rad) -> 2.0 * pi * rad
    | Rectangle(_, _, hgt, wdt) -> 2.0 * (hgt + wdt)

// show_shape : Shape -> unit
// Displays the data of the given shape.
let show_shape shp =
    match shp with
    | Circle(clr, loc, rad) ->
        printfn "Color = %A, Location (%.3f, %.3f)" clr loc.x loc.y
        printfn "Kind = Circle, Radius = %.3f" rad
    | Rectangle(clr, loc, hgt, wdt) ->
        printfn "Color = %A, Location (%.3f, %.3f)" clr loc.x loc.y
        printfn "Kind = Rectangle, Height = %.3f, Width = %.3f" hgt wdt
    
    printfn "Area = %.3f, Perimerter = %.3f\n" 
        (area_shape shp) (perim_shape shp)

// start : unit -> unit
// Serves as an entry point for the program.
let start () =
    let lst = [
        Circle(Red, { x = 1.0; y = 2.0 }, 5.0);
        Rectangle(Green, { x = 10.0; y = 15.0 }, 4.0, 2.0);
        Circle(Blue, { x = 9.0; y = 3.0 }, 6.0)
    ]
    printfn "These are the shapes:"
    List.iter show_shape lst

File: Prolog Workspace/03_Structures/shapes_calc.pl
/*
 * shapes_calc.pl
 * By Antonio F. Huertas
 * This program displays the data, including the area and perimeter,
 * of a list of shapes.
 */

%% pi(?ValuePi)
%  Defines the value of pi.
pi(3.14159).

%% color(?ValueColor)
%  Defines the different colors.
color(white).
color(red).
color(green).
color(blue).
color(yellow).
color(black).

%% area(+Shape, ?Area)
%  Computes the area of the given shape.
area(circle(_, _, Radius), Area) :-
    pi(ValuePi),
    Area is ValuePi * Radius ** 2.0.

area(rectangle(_, _, Height, Width), Area) :-
    Area is Height * Width.

%% perimeter(+Shape, ?Perimeter)
%  Computes the perimeter of the given shape.
perimeter(circle(_, _, Radius), Perimeter) :-
    pi(ValuePi),
    Perimeter is 2 * ValuePi * Radius.

perimeter(rectangle(_, _, Height, Width), Perimeter) :-
    Perimeter is 2 * (Height + Width).

%% write_data(+Shape)
%  Displays the data of the given shape.
write_data(Shape) :-
    Shape = circle(point(X, Y), color(Color), Radius),
    write('\tLocation = '), write(X), write(', '), write(Y),
    write(', Color = '), write(Color), nl,
    write('\tType = circle, Radius = '), write(Radius), nl,
    write_results(Shape), nl.

write_data(Shape) :-
    Shape = rectangle(point(X, Y), color(Color), Height, Width),
    write('\tLocation = '), write(X), write(', '), write(Y),
    write(', Color = '), write(Color), nl,
    write('\tType = rectangle, Height = '), write(Height),
    write(', Width = '), write(Width), nl,
    write_results(Shape), nl.

%% write_results(+Shape)
%  Displays the area and perimeter of the given shape.
write_results(Shape) :-
    area(Shape, Area),
    write('\tArea = '), write(Area),

    perimeter(Shape, Perimeter),
    write(', Perimeter = '), write(Perimeter).

%% write_list(+List)
%  Displays the elements of the given list.
write_list([Head|Tail]) :-
    write_data(Head), nl,
    write_list(Tail).

write_list([]).

%% start
%  Serves as an entry point for the program.
start :-
    List = [
        circle(point(1.0, 2.0), color(red), 5.0),
        rectangle(point(10.0, 15.0), color(green), 4.0, 2.0),
        circle(point(9.0, 3.0), color(blue), 6.0) ],
    write('These are the shapes:\n'),
    write_list(List), nl.

/*
 * Sample interactive run:
 * ?- start.
 */